### PR TITLE
Generate DICE data in Stage 1

### PR DESCRIPTION
**Pull request by conradgrobler**
The main steps involved are:

- Convert the DICE evidence struct from Stage 0 into a proto representation
- Measure the system image and build a new ECA certificate for the next layer
- Save the new DICE data to a file in the system image where the orchestrator can read it

Ref https://github.com/project-oak/oak/issues/4074

### CODE REVIEW

File: oak_containers_stage1/src/dice.rs [hunk 1]
```diff
@@ -44,8 +68,111 @@ const MEMMAP_PATH: &str = "/sys/firmware/memmap";
...
 const PHYS_MEM_PATH: &str = "/dev/mem";
 
+/// Builds the DICE evidence and certificate authority for the next DICE layer.
+pub struct DiceDataBuilder {
+    evidence: Evidence,
+    signing_key: SigningKey,
+    system_image_measurement: Vec<u8>,
+}
+
+impl DiceDataBuilder {
```
Comment (ipetr0v) [inline oak_containers_stage1/src/dice.rs:78]: I think originally we were planning to have some of this logic in the `oak_dice` crate.

I.e. to have something like:
```rust
struct Attester {
  fn create(DiceData dice_data) -> Self;
  fn generate_evidence(Measurements measurements, ApplicationKeys application_keys) -> Evidence;
  fn generate_dice_data() -> DiceData;
}
```
But it could be a separate PR

**Review by ipetr0v — APPROVED** [review]

File: oak_containers_stage1/src/dice.rs (hunk 1 above)
Comment (conradgrobler) [reply oak_containers_stage1/src/dice.rs:78]: Yes, this requires some further design. I started with this approach, but since `oak_dice` is used by Stage 0 and we are trying to avoid protos as a dependency in Stage 0 adding this there is problematic. We would either have to create two dice crates (one with protos and one without) or accept that we will have protobufs as a dependency in Stage 0 and the Restricted Kernel.

**Comment by conradgrobler** [comment]
> what's going on with CBOR?

We have decided to use CBOR-based CWT certificates for the DICE evidence rather than X.509 (those are the two formats supported in the Open DICE profile: https://pigweed.googlesource.com/open-dice/+/HEAD/docs/specification.md#cbor-cdi-certificates)

File: oak_containers_stage1/src/dice.rs (hunk 1 above)
Comment (ipetr0v) [reply oak_containers_stage1/src/dice.rs:78]: The crate with protos could be the `oak_remote_attestation`, which will depend on the `oak_dice`

File: oak_dice/src/cert.rs [hunk 2]
```diff
@@ -47,6 +47,9 @@ pub const MEMORY_MAP_MEASUREMENT_ID: i64 = -4670559;
 /// ID for the CWT private claim ID corresponding to the hash of the commands for building the ACPI
 /// tables.
 pub const ACPI_MEASUREMENT_ID: i64 = -4670560;
+/// ID for the CWT private claim label corresponding to the hash of the binary for the layer in the
+/// case where a single binary is measured for a layer..
+pub const CODE_DIGEST_ID: i64 = -4670545;
```
Comment (souravdasgupta) [inline oak_dice/src/cert.rs:52]: Is this the Layer 1 code digest ID? If not, then where is that ID defined?
If yes, as per [go/tee-dice-attestation](http://goto.google.com/tee-dice-attestation) it seems it should be -4670567.

Comment (conradgrobler) [reply oak_dice/src/cert.rs:52]: Good catch. This is the ID used for generic code measurements in the Open DICE profile, but we agreed to use unique IDs for the different layers. I will update in a follow-up PR.
